### ISSUE TITLE
Random.Extra.frequency doesn't work as specified

### ISSUE THREAD

**Issue by league** [elm-community/elm-random-extra#4]
… or maybe I don't understand how to use it.

I'd like to generate from a specified distribution. The doc says "the likelihood of a given generator being chosen is its likelihood divided by the sum of all likelihood" so giving a list like `[(2,ga), (3,gb), (5,gc)]` would choose `gc` 50% of the time, `gb` 30%, etc. It should be equivalent to `[(0.2,ga), (0.3,gb), (0.5,gc)]`.

But here are some results. Paste this into the `elm repl`:

```
import List.Extra exposing (..)
import Random.Extra exposing (..)
import Random exposing (..)
doit g = List.map List.length <| group <| List.sort <| generateN 1000 g (Random.initialSeed 124)
```

and then we can count the distributions:

```
> doit <| frequency [(2,constant "A"),(3,constant "B"),(5,constant "C")] (constant "D")
[28,25,28,919] : List Int
> doit <| frequency [(0.2,constant "A"),(0.3,constant "B"),(0.5,constant "C")] (constant "D")
[326,352,322] : List Int
```

Using 2,3,5 almost always generates "D", which according to the doc should only be used if the list is empty or the sum of (absolute values of) likelihoods is zero.

Using 0.2, 0.3, 0.5 doesn't work either – it never generates "D", but it generates each of them about a third of the time, instead of favoring "C".

Is this broken, or is the doc wrong? How to call it? Thanks!

**Comment by league** [comment]
Perhaps I'll try to resolve this based on the [similar function](https://hackage.haskell.org/package/QuickCheck-2.8.2/docs/src/Test-QuickCheck-Gen.html#frequency) in Haskell QuickCheck:

``` haskell
-- | Chooses one of the given generators, with a weighted random distribution.
-- The input list must be non-empty.
frequency :: [(Int, Gen a)] -> Gen a
frequency [] = error "QuickCheck.frequency used with empty list"
frequency xs0 = choose (1, tot) >>= (`pick` xs0)
 where
  tot = sum (map fst xs0)

  pick n ((k,x):xs)
    | n <= k    = x
    | otherwise = pick (n-k) xs
  pick _ _  = error "QuickCheck.pick used with empty list"
```

**Comment by mgold** [comment]
This looks very peculiar and I'm investigating. It's possible the statistical issues are do to the known-to-be-bad RNG but you're correct that "D" should never appear.

I personally like Haskell's decision to error on the empty list, since the function will almost always be used with a list literal. But I'm not the original library author.

**Comment by league** [comment]
The other difference is the Elm version uses Float and Haskell uses Int. But I believe it doesn't affect the logic. The `pick` function is very clever.

**Comment by league** [comment]
After applying that patch:

```
> doit <| frequency [(2,constant "A"),(3,constant "B"),(5,constant "C")] (constant "D")
[189,315,496] : List Int
> doit <| frequency [(0.2,constant "A"),(0.3,constant "B"),(0.5,constant "C")] (constant "D")
[189,315,496] : List Int
> doit <| frequency [(8,constant "A"),(1,constant "B"),(1,constant "C")] (constant "D")
[813,95,92] : List Int
```

Yay!

**Comment by mgold** [comment]
Oh boy, I just wrote exactly the same function! Yeah I have no idea what the original author was thinking.

**Comment by mgold** [comment]
Fixed in 1.0.1. Thanks for the patch!
